### ISSUE TITLE
Dense variable constraints are not allowed with sparse gradients

### ISSUE THREAD

**Issue by skearnes** [tensorflow/tensorflow#33755]
I'm training a simple document embedding model with the Keras API in TF 2.0. When I try to add `embeddings_constraint=tf.keras.constraints.UnitNorm(axis=1)` to an `Embedding` layer, I get the following error (truncated):

```
File "C:\Users\Steven\Miniconda3\lib\site-packages\tensorflow_core\python\keras\optimizer_v2\optimizer_v2.py", line 459, in apply_grad_to_update_var
    "Cannot use a constraint function on a sparse variable.")
```

I have verified that the variable is not sparse; the `Embedding` layer creates a dense matrix to hold the embeddings. However, reading [the code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L126) shows that this error is raised when the *gradients* are sparse, which seems normal for embedding layers and should not (AFAIK) affect the application of constraints to the dense variable after the gradient update.

**In summary, variable constraints should be allowed in this context, and it is a bug that they are not.**

**Comment by jvishnuvardhan** [comment]
@skearnes Thanks for reporting this issue. Can you please share a simple standalone code to reproduce the issue? Thanks!

**Comment by RoMa03** [comment]
I have the same problem.  Code based on https://keras.io/layers/embeddings/

    import numpy as np
    from tensorflow.keras.layers import Embedding                         
    from tensorflow.keras.models import Sequential
    from tensorflow.keras.constraints import UnitNorm

    model = Sequential()
    #model.add(Embedding(1000, 64, input_length=10))    # This version works
    model.add(Embedding(1000, 64, input_length=10, embeddings_constraint=UnitNorm))    # This version fails
    input_array = np.random.randint(1000, size=(32, 10))
    model.compile('rmsprop', 'mse')
    output_array = model.predict(input_array)
    assert output_array.shape == (32, 10, 64)

    target = np.zeros_like(output_array)
    model.fit(input_array, target, epochs=5, verbose=1)

**Comment by skearnes** [comment]
+1 thanks @RoMa03

**Comment by Hoiy** [comment]
The error message should come from [optimizer_v2.py line 459](https://github.com/tensorflow/tensorflow/blob/bf9c196f37b9cbb3109b2891aaf9da85bf5f712a/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L459) ?

**Comment by MTDzi** [comment]
Same here.

**Comment by marson666** [comment]
I also meet the bug, is there anyone who solved this? Thx~

**Comment by noamgot** [comment]
Same here...

**Comment by tlimbacher** [comment]
Same here.

**Comment by tzachar** [comment]
Same here

**Comment by osasseville** [comment]
Same here

**Comment by tanzhenyu** [comment]
IIUC this is a problem in both tf v1 optimizer and tf v2 keras optimizer?

**Comment by PingThePeng** [comment]
same here

**Comment by rubjim** [comment]
same problem here

**Comment by SoccoCMOS** [comment]
Encountered the same bug when using tf.gather on a custom layer weights.

**Comment by jianghaoyuan1994** [comment]
same here

**Comment by hussainkarafallah** [comment]
same here, however with some other implementation i was able to do it
check this please
`
class BigClamModel(keras.Model):
  def __init__(self , nodes , communities):
      super(BigClamModel , self).__init__()
      initializer = HalfNormal(0.3 , 0.1)
      self.embedding = keras.layers.Embedding(input_dim=nodes , output_dim = communities  , embeddings_initializer= initializer , name = "embedding" , embeddings_constraint=tf.keras.constraints.UnitNorm(axis=1))
      self.DotLayer = keras.layers.Dot(axes = 1 , name = "dot_product")
      self.out = keras.layers.Lambda(lambda x : (1.0 - keras.activations.exponential(-1.0 * x)) , name = "output")

  def call(self , inputs):
    inp1 , inp2 = inputs[0] , inputs[1]
    U = self.embedding(inp1)
    V = self.embedding(inp2)
    ret = self.out(self.DotLayer([U , V]))
    return ret 
  
  def model(self):
      inp1 = keras.layers.Input(shape = () ,name = "first_endpoints" , dtype = tf.int32)
      inp2 = keras.layers.Input(shape = () ,name = "second_endpoints" , dtype = tf.int32)
      return keras.Model(inputs=[inp1,inp2], outputs=self.call([inp1,inp2]))

`

**Comment by gkiril** [comment]
I have the same problem. Any suggested solutions?

**Comment by tanzhenyu** [comment]
Hey all -- I am still working on this, should be able to submit a change without requiring any user code change.

**Comment by rzilleruelo** [comment]
[this](https://stackoverflow.com/questions/63122880/tensorflow-2-runtimeerror-cannot-use-a-constraint-function-on-a-sparse-variabl) is a simple workaround for the issue.
```python
from tensorflow.keras.layers import Embedding
from tensorflow.keras.constraints import UnitNorm
. . .
emb = Embedding(input_dim, output_dim, name='embedding_name')
norm_layer = UnitNorm(axis=1)
norm_embedding = norm_layer(emb(embedding_id_input))
. . .
```

**Comment by umbra-scientia** [comment]
This is a serious problem and makes tf.keras unsuitable for prototyping modern language models. We're used the same workaround as rzilleruelo, but it damages the optimization trajectory and stability.

Consider an embedding vector with large norm: We want to make larger gradient updates to it, to balance the representation learning, but when normalizing as a layer instead of a constraint we actually scale it down to target norm, and so we end up making *smaller* updates to it instead of larger ones. The problem is even worse with very short embedding vectors, where the value (and therefore the gradients) get scaled up and then the optimizer clobbers the tiny vector with a large update. Unlike some companies, we don't have extra millions to burn on compute as a workaround.

So our workaround for the problems of the workaround, is using a penalty on the logarithm of the computed scaling factor like so:
```python
def PostUnitNorm(v, regularizer=0):
	vlen = tf.norm(v, ord=2, axis=-1, keepdims=True)
	vlen = tf.math.maximum(vlen, 1e-8)
	if regularizer > 0:
		vlen += l.Activation(tf.math.log, activity_regularizer=tf.keras.regularizers.l1(regularizer))(vlen) * 1e-16
	return v / vlen
```
But this is still broken, because the penalty must be obscenely strong to achieve the desired effect, which makes the term able to completely dominate the optimization and halt progress until the embedding vectors come back into the desired scale.

**Comment by jjanousek** [comment]
I have the same problem with tensorflow 2.4.1. Has anyone found another solution that does not involve switching to a functional model?

**Comment by RoMa03** [comment]
Ricardo Zilleruelo's solution of following the embedding layer with a
separate normalisation layer looks pragmatic to me?

from tensorflow.keras.layers import Embeddingfrom
tensorflow.keras.constraints import UnitNorm
. . .emb = Embedding(input_dim, output_dim,
name='embedding_name')norm_layer = UnitNorm(axis=1)norm_embedding =
norm_layer(emb(embedding_id_input))
. . .

He chose to wrap the two as a function but you could easily have them as
separate layers if you prefer.

Robert

**Comment by amit-agni** [comment]
I am getting the same error for non_neg() constraint and the workaround suggested by @rzilleruelo works for me. Now, I am interested in extracting the weights after applying the non_neg() constraint, how can I do that ?

For the Embedded layer I could do this :
model.get_layer(name='embedding_name').get_weights()

**Comment by mohantym** [comment]
I tried to run the code on colab with TF v2.5  after following  [this article](https://github.com/keras-team/keras/issues/14459) and didn't get error after removing this constraint `embeddings_constraint=UnitNorm `in the code .Please find the [gist here](https://colab.research.google.com/gist/mohantym/eb370120505877a14669c6d4eff3cb28/33755.ipynb). Thanks !

**Comment by jeisinge** [comment]
As of TF 2.4, I too am getting this error; I have had to remove embedding constraints.

**Comment by jeisinge** [comment]
Since the defect still exists, I believe it should remain open.

**Comment by jeisinge** [comment]
Nothing has changed. The warnings persist.

**Comment by davidsc-unity3d** [comment]
I just ran into this - vote for a fix - the workaround

```
norm_embedding = norm_layer(emb(embedding_id_input))
```

probably solves must use cases, but it is essentially a constraint on the activity - the constraint is applied after the embeddings are looked up. The constraint one passes to the keras Embedding, per the documentation, is on the full matrix (like the embeddings_regularizer).

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!

**Comment by davidsc-unity3d** [comment]
I added an alternative workaround to https://stackoverflow.com/a/70427151/2280020, @umbra-scientia

**Comment by pjpratik** [comment]
@skearnes Could you please check at this [gist](https://colab.research.google.com/gist/pjpratik/5735a80d0f6f89b2bafd3f226455d5e3/33755.ipynb) for reference and let us know if it helps. Thank you.
